hgtv/views/playlist.py

```diff
@@ -230,45 +271,50 @@ def playlist_import(channel):
             db.session.add(playlist)
             db.session.commit()
             cache.delete('data/featured-channels')
+            return make_response(jsonify(status='ok', doc=_(u"Imported playlist {title}.".format(title=playlist.title)), result={'new_playlist_url': playlist.url_for()}), 201)
         except (DataProcessingError, ValueError) as e:
-            flash(e.message, category="error")
-            return render_form(form=form, title="Import Playlist", submit=u"Import",
-                cancel_url=channel.url_for(), ajax=False)
-        flash(u"Imported playlist '%s'." % playlist.title, 'success')
-        return render_redirect(playlist.url_for(), code=303)
-    return render_form(form=form, title="Import Playlist", submit=u"Import",
-        cancel_url=channel.url_for(), ajax=False)
+            return make_response(jsonify(status='error', errors={'error': [e.message]}), 400)
+    return make_response(jsonify(status='error', errors=form.errors), 400)
+
+
+@app.route('/<channel>/import', methods=['GET', 'POST'])
+@lastuser.requires_login
+@render_with({'text/html': 'index.html.jinja2', 'application/json': handle_import_playlist})
+@load_model(Channel, {'name': 'channel'}, 'channel', permission='new-playlist')
+def playlist_import(channel):
+    return dict(channel=channel)
+
+
+def handle_playlist_extend(data):
+    channel = data['channel']
+    playlist = data['playlist']
+    form = PlaylistImportForm()
+    form.channel = channel
+    if request.method == 'GET':
+        html_form = render_form(form=form, title=u"Playlist extend", submit=u"Save",
+        cancel_url=playlist.url_for(), ajax=False, with_chrome=False)
+        return jsonify(playlist=get_playlist_details(channel, playlist, videos_count='none'), form=html_form)
+    if form.validate_on_submit():
+        playlist_url = escape(form.playlist_url.data)
+        initial_count = len(playlist.videos)
+        try:
+            process_playlist(playlist_url=playlist_url, playlist=playlist)
+        except:
+            return make_response(jsonify(status='error', errors={'error': ['Oops, something went wrong, please try later']}), 400)
+        additions = (len(playlist.videos) - initial_count)
+        if additions:
+            db.session.commit()
+            cache.delete('data/featured-channels')
+        return make_response(jsonify(status='ok', doc=_(u"Added video to playlist {title}.".format(title=playlist.title)), result={}), 200)
+    return make_response(jsonify(status='error', errors=form.errors), 400)
 
 
 @app.route('/<channel>/<playlist>/extend', methods=['GET', 'POST'])
 @lastuser.requires_login
+@render_with({'text/html': 'index.html.jinja2', 'application/json': handle_playlist_extend})
```

Comment:
@jace is it fine to move out view code in an outer function like this? I haven't done this yet, so not sure.